### PR TITLE
Validation model specs n+1 problem

### PR DESCRIPTION
**Pull request by Somdotta07**
### **Project requirements**
- Added the following validations:
  - For the User model:
    - Name must not be blank.
    - PostsCounter must be an integer greater than or equal to zero.
  - For the Post model:
    - Title must not be blank.
    - Title must not exceed 250 characters.
    - CommentsCounter must be an integer greater than or equal to zero.
    - LikesCounter must be an integer greater than or equal to zero.
- Added unit specs for all of the models' methods and validations.
- Added flash messages in the create actions in all controllers.
- Solved the N+1 problem when fetching all posts and their comments for a user.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -22,9 +22,10 @@ def create
     respond_to do |format|
       format.html do
         if @new_post.save
-          redirect_to "/users/#{@new_post.user.id}/posts/", notice: 'Success!'
+          redirect_to "/users/#{@new_post.user.id}/posts/", flash[:notice] = 'Your post is saved!'
```
Comment (Dipeshtwis) [inline app/controllers/posts_controller.rb]: - [x] While creating a post, I'm getting the error of `no implicit conversion of symbol into string`. Please fix this.

Screenshot from 2022-02-25 22-56-38

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -13,9 +13,10 @@ def create
     @new_comment.post_id = @post.id
     @new_comment.update_comments_counter
     if @new_comment.save
-      redirect_to "/users/#{@post.user_id}/posts/#{@post.id}", notice: 'Success!'
+      redirect_to "/users/#{@post.user_id}/posts/#{@post.id}", flash[:notice] = 'Your comment is saved!'
```
Comment (Dipeshtwis) [inline app/controllers/comments_controller.rb]: - [x] Same here. While creating a comment, I'm getting the error of `no implicit conversion of symbol into string`. Please fix this.

Screenshot from 2022-02-25 22-59-07

File: app/controllers/likes_controller.rb [hunk 3]
```diff
@@ -7,9 +7,9 @@ def create
     )
     new_like.update_likes_counter
     if new_like.save
-      redirect_to "/users/#{@post.user_id}/posts/#{@post.id}", notice: 'Success!'
+      redirect_to "/users/#{@post.user_id}/posts/#{@post.id}", flash[:notice] = 'Your Like is saved!'
     else
-      redirect_to "/users/#{@post.user_id}/posts/#{@post.id}", alert: 'Error occured!'
+      redirect_to "/users/#{@post.user_id}/posts/#{@post.id}", flash.now[:error] = 'Could not save like'
```
Comment (Dipeshtwis) [inline app/controllers/likes_controller.rb]: - [x] Same here. While creating a like, I'm getting the error of `no implicit conversion of symbol into string`. Please fix this.

Screenshot from 2022-02-25 22-58-35

File: spec/models/like_spec.rb [hunk 4]
```diff
@@ -1,5 +1,19 @@
...
+    like.build_user(name: 'Somdotta')
+    like.build_post(text: '')
+    expect(like).to be_valid
+  end
+
+  it 'updates a posts likes correctly' do
+    user = User.new(name: 'Tom', posts_counter: 0)
+    user2 = User.create!(name: 'Lilly', posts_counter: 0)
+    post = user2.posts.create!(title: 'Post1', text: 'This is a post', likes_counter: 0, comments_counter: 0)
+    post.likes.create!(user:)
```
Comment (Dipeshtwis) [inline spec/models/like_spec.rb:15]: - [x] On line no. 15, I think you have missed mentioning user in the parameter. It should be like:

`post.likes.create!(user: user)`.

**Review by Dipeshtwis — CHANGES_REQUESTED** [review]
Hi @Somdotta07, 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


 ## Required Changes ♻️

**_Check the comments under the review_**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

- N.A.

Cheers and Happy coding!👏👏👏
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**_Dipesh Kumar_**


------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project(check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (Somdotta07) [reply app/controllers/posts_controller.rb]: Thanks for your time and feedback. I'll fix it ASAP.

File: app/controllers/comments_controller.rb (hunk 2 above)
Comment (Somdotta07) [reply app/controllers/comments_controller.rb]: Thanks for your time and feedback. I'll fix it ASAP.

File: app/controllers/likes_controller.rb (hunk 3 above)
Comment (Somdotta07) [reply app/controllers/likes_controller.rb]: Thanks for your time and feedback. I'll fix it ASAP.

File: spec/models/like_spec.rb (hunk 4 above)
Comment (Somdotta07) [reply spec/models/like_spec.rb:15]: Thanks for your time and feedback. I'll fix it ASAP. I remember I had to remove `user` to fix Rubocop

**Comment by Somdotta07** [comment]
Fixed all the issues shown by the previous reviewer, but  I had to remove` user` to fix Rubocop

**Review by descholar-ceo — APPROVED** [review]
Hi @Somdotta07 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**